### DISCUSSION TITLE
Does LUA know which modulator did the call?

### DISCUSSION THREAD

**Discussion by bijlevel** [RomanKubiak/ctrlr#154] [Q&A]
I have several buttons calling each time the same LUA function when pressed. Does LUA know which button/modulator made the call? If so, how does that look in LUA speak? TIA!

Sjoerd W

**Comment by Ted-juh** [comment]
Let's say you attached a method to luaModulatorValueChange of two buttons,  button1 and button2, and you want to know which button is which:

buttonNameOfButtonThatIsPressed = function(mod, value)

-- Press a button, the value changed from 0 to 1 or vice versa, Ctrlr passes the value change on to the method.
-- The method will only act as a listener to the buttons that have the method attached.

-- get the name of the modulator of which the value changed
name = L(mod:getName())
-- show it in the console
console("name"..name) 

-- Lua knows the name of the modulator which value changed, now do something with it:
if name == "button1" then
console("button1doingsomecrazyassfunkystuff")
else
console("button2doingsomemorecrazyshizzlefizzle")
end
end

Does that answer your question?

**Reply by bijlevel** [reply]
Thanks a lot for the quick answer! I tried it in several ways, but I probably do something wrong, because I get an "nil" error on the "mod" value. I will try to explain better what I'm after.

I have a row of 8 buttons (on/off). Each button sends a different Sysex formula, but calls the same method (in my case called "getLCDText()") to retrieve some text to display (text differs depending on which button is pressed).

In the panel layout, besides each button is an image that mimics a burning LED. As soon as the method is called I want the LED image to be visible so it gives a visual feedback of which button is pressed (and set the LED of a former button to invisible, so I need to go through the remaining 7 LED's to know which of them is visible to set that one to invisible). (I know I can do this all by changing the appearance of the button when pressed, but this is how the hardware machine for which I'm making this panel, does it.) This is how I do it for one (known) button:

**function getLCDText()
	showLed() -- this calls the function to show the LED image
	<some code follows, irrelevant for this discussion>
end**

So I'm calling the function **showLED()**. This is the function:

**function showLed()
    panel:getModulatorByName("LED1"):getComponent():setProperty("componentVisibility", 1, false)
end**

This works of course for the hard coded name "LED1". But to be more general, "LED1" needs to be replaced with e.g. a variable 'name' which holds the name of the LED image that needs to be visible. If I know the name of the button pressed I know the name of the corresponding LED image.

Sorry for this extensive answer, but I hope it explains better what I'm after.

KR Sjoerd

**Reply by Ted-juh** [reply]
Ok. I get it. But before we go through the Lua code that is needed to add something like that, why not make a button in Photoshop or something that's basically the button and the LED next to each other? In that case you need uiImageButton to do it all without a single line of code.

**Reply by bijlevel** [reply]
Good advise, but the button needs to be visible all the time, while the LED should be switchable on or off, so visible/invisible. I can use two images for the button/LED combo, one with the LED on and one with the LED off, but while I'm writing this, I think this makes it more complex. I will first try the code option(s).

**Comment by dobo365** [comment]
Try using:
panel:getModulatorByName("LED1"):getComponent():setVisible(true)
or false of course

To manage several LEDs with one function, you use the method described above
sName = L(mod:getName())
panel:getModulatorByName(sName):getComponent():setVisible(true)

**Reply by Ted-juh** [reply]
Wouldn't that set the visibility of the button that is pressed itself, Dobo?

Quick mockup:
[buttonTest.zip](https://github.com/RomanKubiak/ctrlr/files/5939149/buttonTest.zip)

**Reply by dobo365** [reply]
Yes indeed. I was just giving the code to use. 
To be more precise:

```
sName = L(mod:getName())
if sName == "LED1" then
panel:getModulatorByName("whatever_Button1"):getComponent():setVisible(true)
-- do something else
elseif sName == "LED2" then
panel:getModulatorByName("whatever_Button2"):getComponent():setVisible(true)
-- do something else
end
```

Then you can extend as you want. Capture the value of the LED to set visible or not, do something else...
You should also avoid using panel:getModulatorByName() in all methods. Better to do it once in a main method by assigning variable then use those in the different methods.
```
modMod1 = panel:getModulatorByName("Mod1")
modMod2 = panel:getModulatorByName("Mod2")
```
then use
`modMod1:getComponent():setVisible(true)`
 in your methods

But that, I have already explained 100 times ;-)

**Reply by bijlevel** [reply]
Hi,
Thanks a lot for the input! Very much appreciated. Didn't have time yet to try it out. I will try it this evening and will of course let you know if it works!

**Reply by Ted-juh** [reply]
Bijlevel, you got two sets of code to work with. 

- Dobo uses an if/ else statement. 
- I uploaded a panel (button test.zip) in my last post, it uses a method without an if/ else statement. But for readability's sake, I didn't do what Dobo said about declaring your modulators in a "main" method. Although that's good practice.

Good luck,  and yes, always report back.

**Reply by bijlevel** [reply]
Hi,
I did use the code from the uploaded panel 'button test' by DnaldeT (zéér bedankt!) and got it working with a few modifications.

`function getLCDText(mod)
	butIndex = L(mod:getProperty("modulatorCustomIndex"))
	-- returns modulatorCustomIndex of the button
	-- that is "if" you give the modulatorCustomIndex of the (in this case) 8 buttons a value of 1 to 8
	-- ModulatorCustomIndex is a few properties up from name in the properties pane.

	sName = string.format("LED%s",butIndex)
	-- returns LED1, LED2 etc

	panel:getModulatorByName(sName):getComponent():setProperty("componentVisibility", 1, false) -- turn the LED on
	
	-- next action: dim the previous LED. For this we need to check all 8 buttons
	for i=1,8 do
		hName = string.format("LED%s",i)
		console(invName)
		if hName ~= sName then
			panel:getModulatorByName(hName):getComponent():setProperty("componentVisibility", 0, false) --turn LED off
		end
	end
	...
end`

So thanks everybody for the input. @dobo365: I did not (yet) declare my modulators in a "main" method, because this a relative small panel (until now). But I see the wisdom in this approach and will make it a good habit of me :-)!

BTW odd enough GitHub did not format all of my code...

**Reply by Ted-juh** [reply]
Good to see it works now. Don't forget to set the topic to "answered". 

About the formatting of code. For a platform that's about coding, the formatting of code is very dodgy, buggy indeed.

**Comment by dobo365** [comment]
OK. I repeat it for the 101th time (you can search for this in the old forum):

1/ Create a method called AssignModulators() like:
```
function AssignModulators()

	-- Main Screen
	txtLCD_Program = panel:getModulatorByName("LCD_Program")
	txtLCD_ProgramName = panel:getModulatorByName("LCD_ProgramName")

	-- Program
	modProgramLevel = panel:getModulatorByName("ProgramLevel")
	modProgramTranspose = panel:getModulatorByName("ProgramTranspose")
	modProgramTuning = panel:getModulatorByName("ProgramTuning")

	-- Oscillators 
	modVCO1Wave = panel:getModulatorByName("VCO1Wave")
	modVCO1Octave = panel:getModulatorByName("VCO1Octave")
	modVCO1Pitch = panel:getModulatorByName("VCO1Pitch")

end
```

2/ Create a PanelLoaded() method that you attach to the panel property "Called when the panel ahs finish loading"
It will execute the AssignModulators() method but you can add different things inside it
```
--
-- Called when the panel has finished loading
--
PanelLoaded = function()

	AssignModulators()
	Tables_Init()
	BasicProgramData_Init()
	About_Init()

end
```

3/ good luck ;-)

/goodweather

**Reply by bijlevel** [reply]
I did put your advice already to good use. I can use it as a start up method and indeed avoid having to use some method over and over again! Thanks for reminding me, because way back I did already use this, but forgot about it. BTW about my initial issue, after spending many hours finding out why some simple coding didn't work, I found out that the variable (butIndex) DnaldeT uses (the custom assigned index of a modulator) in his 'button panel' example, turns out to be a string instead of a number! Probably many of you know this, but I didn't and it drove me crazy. How can one know this?

**Reply by Ted-juh** [reply]
There's probably someone who has a better explanation but the short answer is "by trial and error". 

The 1 in led1 is a string indeed although it's a number. But it.comes from the custom modulator index which is a string itself, if I'm correct. I know that because else it would show an error and I would have to use tonumber(butIndex) to convert it to a number. The other way round would be tostring(butIndex) to convert a number to string. But since we already take a string, it's not necessary to do that conversion.

Again, that's the way I understood it but might be completely off.

**Reply by bijlevel** [reply]
Thanks for the reply. Yes, indeed trial and error. I usually know when something is a string, number or whatever, but in this case I assumed that an index is a number, because it is so in the non-programming world and also in many program environments...

**Reply by Ted-juh** [reply]
Well, I think there is a way to find out whether you are taking a string or number. Surf back to the Ctrlr Github Code page and go to: Source/Resources/XML and open CtrlrIDs.XML with notepad++ or whatever suits you. There you'll find all the options of the "panel mode" properties sorted by component/ category. 

Take a look at what it says after "Type". If it says numeric, it takes a number, and vice versa if it says "Text" it takes a string. So, whenever you are in doubt or to exclude a coding mistake, take a look over there.

**Reply by bijlevel** [reply]
Thanks again! Very nice list, I will try to memorise it :-) I remember now that such lists exist, but I recently took up Ctrlr again after 4 years and obviously I have forgotten all the information that there is, be it scattered.

**Comment by dnaldoog** [comment]
@bijlevel 

You might find this code useful.

Also with regard to formatting code, I am using html &lt;pre&gt;&lt;/pre&gt; tags. Seems to work better!

<pre>
programSelect = function(mod,value,source)
    if source == 4 then
        local sName = L(mod:getName())
        local t = {"PROGRAM01", "PROGRAM02", "PROGRAM03", "PROGRAM04"} -- local table of 4 radio buttons
        for i, v in pairs(t) do
            if sName == v then -- run code on clicked on button
                local sysMess = string.format("F0 00 00 23 01 43 C1 00 %.2X F7", i - 1)
                panel:sendMidiMessageNow(CtrlrMidiMessage(sysMess))
            else -- run code on the other buttons that weren't clicked on
                _G[v]:setValue(0, true) 
               -- don't use panel:getComponent(v) // use panel:getModulatorByName(v)
            end
        end
    end
end -- source == 4 (user clicked on a button - not lua generated - that would be 5 or 6) 

--[[ 
*NOTE: Where PROGRAM01 etc are initialised in an init script
"Called when the panel has finished loading"
PROGRAM01=panel:getModulatorByName("PROGRAM01")
PROGRAM02=panel:getModulatorByName("PROGRAM02")
PROGRAM03=panel:getModulatorByName("PROGRAM03")
PROGRAM04=panel:getModulatorByName("PROGRAM04")
--]]
</pre>

https://ctrlr.org/radio-button-code-generator/

**Reply by bijlevel** [reply]
Thank you for your input, much appreciated, but I did solve it (kind off in my way) with the help of the "button panel" example. But your code may come in handy someday.

**Reply by Ted-juh** [reply]
> -- don't use panel:getComponent(v) // use panel:getModulatorByName(v)

Now, this is what still gives me headaches from time to time.

As explained by Goodweather in the "Panel: General, I don't get it" topic on the old forums: A Component is a child of its parent, the Modulator. Does that mean a Component inherits the name of its Modulator? Because panel:getModulatorByName(k):getComponent():setValue() and its alternative panel:getComponent(k): setValue() do suggest so. Then why is it that you can't use (or shouldn't use) panel:getComponent() in your example?

**Reply by dobo365** [reply]
('Im goodweather but it doesn't want to show my nickname ;-) )

You have to see it indeed like that:
- a modulator is an object and has properties and has methods that can be applied to it
- a component is a child object of the modulator and has properties that can be applied to it
- the Midi part is another child object of the Modulator (MidiMessage)
- you cannot use the methods of a component on a modulator or the opposite
- a panel is another object as well and so on...

To reach a modulator, you would do (bad method to start from the panel but good example)
panel:getModulatorByName("modulator-1")
To reach the component of modulator-1 you do
panel:getModulatorByName("modulator-1"):getComponent()
To reach the Midi message properties under the modulator-1 you do
panel:getModulatorByName("modulator-1"):getMidiMessage()

You find the properties of a modulator by looking on the Property panel under Modulator
You find the properties of a component within a modulator by looking to the Property panel under Component
You find the properties of a Midi message within a modulator by looking to the Property panel under Midi
You can see the properties name by switching View- Property IDs/Name in the menu (you need to select another mod to see the effect)
Then you can modify them by using getProperty() and setProperty() methods that you apply on the object
panel:getModulatorByName("modulator-1"):getComponent():setProperty("componentLabelHeight", 20, true)
panel:getModulatorByName("modulator-1"):getMidiMessage()():getProperty("midiMessageCtrlrNumber")

To list the methods of a modulator you create one component uiSlider for example (yes, this is here that is confusing!!! You ask toi create a component but you get a modulator Hahaha...)
In fact it creates a modulator that contains your uiSlider component
Suppose it is named modulator-1
In the console, you type 
what(panel:getModulatorByName("modulator-1"))
and you get the modulator methods
If you type
what(panel:getModulatorByName("modulator-1"):getComponent()) you will get another list (much longer)

Hope this will support your understanding ;-)

**Reply by Ted-juh** [reply]
Great explanation and it definitely extends my understanding/ knowledge on how to reach certain properties of a modulator or  one of its components.

**Comment by dobo365** [comment]
and also keep in mind that you can always combine strings and numbers without using the format function but by using the tostring() function:
```
sNewYear = "Hello "..tostring(2021)
console(sNewYear)
```
or more in general
```
iPatchNumber = 100
console(tostring(iPatchNumber))
```
